From bbd14a5ed0c61a3db1f6e9376bc87031068970c2 Mon Sep 17 00:00:00 2001
From: Yabin Cui <yabinc@google.com>
Date: Thu, 25 May 2023 18:08:21 +0000
Subject: [PATCH] [libc++][Android] Disable test_no_resolve_symlink_on_symlink
 on Android

The test is intended to ensure a particular behavior, but different
versions of Android (even just L and up) have multiple possible
behaviors.

Starting with Android M, fchmodat appears to successfully modify the
permissions of a symlink on an ext4 filesystem (but not f2fs). With
Android R, fchmodat then fails with operation_not_supported, but still
modifies the symlink's permissions. This issue is tracked internally at
http://b/262631136.

Differential Revision: https://reviews.llvm.org/D140183

Change-Id: I2e091313031e080c71a5821070c8ffab50dacc76
---
 .../fs.op.permissions/permissions.pass.cpp      | 17 +++++++++++++++--
 1 file changed, 15 insertions(+), 2 deletions(-)

diff --git a/libcxx/test/std/input.output/filesystems/fs.op.funcs/fs.op.permissions/permissions.pass.cpp b/libcxx/test/std/input.output/filesystems/fs.op.funcs/fs.op.permissions/permissions.pass.cpp
index dc5ff1188521..4b6aff7729b3 100644
--- a/libcxx/test/std/input.output/filesystems/fs.op.funcs/fs.op.permissions/permissions.pass.cpp
+++ b/libcxx/test/std/input.output/filesystems/fs.op.funcs/fs.op.permissions/permissions.pass.cpp
@@ -135,10 +135,23 @@ static void basic_permissions_test()
     }
 }
 
-#ifndef _WIN32
+#if !defined(_WIN32) && !defined(__ANDROID__)
 // This test isn't currently meaningful on Windows; the Windows file
 // permissions visible via std::filesystem doesn't show any difference
 // between owner/group/others.
+//
+// Skip the test on Android because the behavior varies widely. A device
+// may do one of these things:
+//  - Ignore nofollow and modify the target (e.g. L emulator).
+//  - Refuse to let the shell user create a symlink (e.g. physical L device,
+//    volantis or Nexus 7 2013).
+//  - Change the symlink's permissions successfully (e.g. emulators from
+//    Android M to Q, ext4 filesystem).
+//  - Change the symlink's permissions, then fail with operation_not_supported
+//    anyway (e.g. R emulator and later, ext4 filesystem, see Google internal
+//    bug http://b/262631136).
+//  - Do nothing and fail with operation_not_supported (e.g. devices using f2fs
+//    for /data/local/tmp).
 static void test_no_resolve_symlink_on_symlink()
 {
     scoped_test_env env;
@@ -187,7 +200,7 @@ int main(int, char**) {
     test_signatures();
     test_error_reporting();
     basic_permissions_test();
-#ifndef _WIN32
+#if !defined(_WIN32) && !defined(__ANDROID__)
     test_no_resolve_symlink_on_symlink();
 #endif
     return 0;
-- 
2.41.0.rc0.172.g3f132b7071-goog

